src/pg_copy.c

```diff
@@ -0,0 +1,626 @@
+					
+					/* write the row to the shard */
+					if (nextRowFound)
+					{		
+						if (columnNulls[partitionColumn->varattno]) { 
+							ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+											errmsg("cannot copy row with NULL value "
+												   "in partition column")));
+						} else { 
+							Datum partitionValue = columnValues[partitionColumn->varattno];
+							rightConst->constvalue = partitionValue;
+							prunedList = PruneShardList(tableId, whereClauseList, shardIntervalList);
+							
+							foreach(shardIntervalCell, prunedList)
+							{
+								ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+								int64 shardId = shardInterval->id;
+								bool found;
+								
+								copyConn = (CopyConnection*)hash_search(shard2conn, &shardInterval->id, HASH_ENTER, &found);
+								if (!found) { 
+									List *finalizedPlacementList;
+									copyConn->nReplicas = 0;
+									copyConn->preparedMask = 0;
+									/* grab shared metadata lock to stop concurrent placement additions */
+									LockShardDistributionMetadata(shardId, ShareLock);
+									
+									/* now safe to populate placement list */
+									finalizedPlacementList = LoadFinalizedShardPlacementList(shardId);
+									foreach(taskPlacementCell, finalizedPlacementList)
+									{
+										ShardPlacement *taskPlacement = (ShardPlacement *) lfirst(taskPlacementCell);
+										char *nodeName = taskPlacement->nodeName;
+										char nodePort[16];
+										PGconn* conn;
+										sprintf(nodePort, "%d", taskPlacement->nodePort);	 
+										conn = ConnectToNode(nodeName, nodePort);
+										if (conn != NULL)										 
+										{
+											copyConn->conn[copyConn->nReplicas++] = conn;
+											/*
+											 * New connection: start transaction with copy command on it.
+											 * Append shard id to table name.
+											 */
+											if (!tmgr->Begin(conn) ||
+												!PgShardExecute(conn, PGRES_COPY_IN, "%.*s_%ld%s", relationSuffixPos, query,
+																(long)shardId, query + relationSuffixPos))
+											{
+												continue;
+											}
+										} else { 
+											elog(ERROR, "Failed to connect to node %s:%s", nodeName, nodePort);
+										}
+									}
+								}
+								/* TODO: handle binary format */
+								lineBuf = CopyGetLineBuf(copyState);
+								lineBuf->data[lineBuf->len++] = '\n'; /* there was already new line in the buffer, but it was truncated: no need to heck available space */
+								/* Replicate row to all shards */
+								for (i = 0; i < copyConn->nReplicas; i++) { 
+									PQputCopyData(copyConn->conn[i], lineBuf->data, lineBuf->len);
+								}
+							}
+						}
+					}
+					MemoryContextReset(tupleContext);
+				}
+				/* Perform two phase commit in replicas */
+				hash_seq_init(&hashCursor, shard2conn);
+				while (failedShard < 0 && (copyConn = (CopyConnection*)hash_seq_search(&hashCursor)) != NULL) {
```

Comment:
Sorry, I do not understand completely your idea. What should be linked in copyStreamList?
Frankly speaking, I do not think that it can simplify the code or make it more reusable...
May be I am wrong, because I do not understand what you mean.